File: SimulationJava/src/Zwierze.java
import java.util.Random;

public abstract class Zwierze  extends Organizm {


    protected void ruch(noweXY nowe){
        Random rand = new Random();
        int g=0,gdzie;
        int[][]t=s.getSasiad(this);
        int[] t2 = new int[s.getIleS()];
        for(int i=0;i<s.getIleS();i++){
            if(t[i][xpos]!=-1){
                 t2[g]=i;
                 g++;
            }  
        }
        if(g!=0){
            gdzie = rand.nextInt(g);
            nowe.nX = t[t2[gdzie]][xpos];
            nowe.nY = t[t2[gdzie]][ypos];
        }
    }

    protected void tura(noweXY nowe){
        if(s.getPole(nowe.nX, nowe.nY)==null){
            s.setTab(nowe.nX, nowe.nY, this);
            s.setTab(x, y, null);
            this.x=nowe.nX;
            this.y=nowe.nY;
        } else {
            if(s.getPole(nowe.nX, nowe.nY).wypiszNazwe()==wypiszNazwe()){
                if(plodnosc==zdolnydoreprodukcji && s.getPole(nowe.nX, nowe.nY).getPlodnosc()== zdolnydoreprodukcji && this != s.getPole(nowe.nX, nowe.nY)){
                    s.getPole(nowe.nX, nowe.nY).setPlodnosc(nieplodny);
                    s.getPole(nowe.nX, nowe.nY).setZycie(mlody);
                    plodnosc = nieplodny;
                    Organizm n = dziecko();
                    reprodukcja(nowe.nX,nowe.nY, n);
                }
            } else {
                s.getPole(nowe.nX, nowe.nY).kolizja(x,y);
            } 
        }
    }

    public void reprodukcja (int xx, int yy, Organizm n){
        n.setZycie(mlody);
        n.setPlodnosc(plodny);
        boolean miejsce = false;
        for(int z =0 ; z < 2 ; z++){
            int[][] t=s.getSasiad(this);
            if(!miejsce){
                if (z == 1) t=s.getSasiad(s.getPole(xx, yy));
                for(int i=0;i<s.getIleS();i++){
                   if(t[i][xpos]!=BrakSasiada){
                       if(s.getPole(t[i][xpos], t[i][ypos])==null){
                           miejsce=true;
                           n.setX(t[i][xpos]);
                           n.setY(t[i][ypos]);
                           s.dodaj(t[i][xpos], t[i][ypos], n);
                           makeLog(wypiszNazwe(), wypiszNazwe(), 1, n.getX(), n.getY());
                           return;
                       }
                   }
                }
            }
        }
    }

    public void makeLog(String A, String B, int co,int x,int y){
        if(co==1) s.setLog(s.getLog() + A +" rodzi sie na: (" + x + "," + y + ")" + "\n");
        if(co==2) s.setLog(s.getLog() + A +" zabija "+ B +" na: (" + x + "," + y + ")" + "\n");
        if(co==3) s.setLog(s.getLog() + A +" atakuje " + "\n");
        if(co==4) s.setLog(s.getLog() + A +" odbija "+ B +" na: (" + x + "," + y + ")" + "\n");
        if(co==5) s.setLog(s.getLog() + A +" ucieka od "+ B +" na: (" + x + "," + y + ")" + "\n");
        if(co==6) s.setLog(s.getLog() + A +" nie walczy z "+ B +" na: (" + x + "," + y + ")" + "\n");
    }

    @Override
    protected void kolizja(int xx,int yy) {
        makeLog(s.getPole(xx, yy).wypiszNazwe(), "", 3, x, y);
        if(s.getPole(xx, yy).getSila() >= sila){
            zycie=0;
            s.setTab(x, y, s.getPole(xx, yy));
            s.setTab(xx, yy, null);
            s.getPole(x, y).setX(x);
            s.getPole(x, y).setY(y);
            makeLog(s.getPole(x, y).wypiszNazwe(), wypiszNazwe(), 2, x, y);
        } else {
            if(s.getPole(xx, yy).getSila() < sila){
                makeLog(wypiszNazwe(), s.getPole(xx, yy).wypiszNazwe(), 2, x, y);
                s.getPole(xx, yy).setZycie(martwy);
                s.setTab(xx, yy, null);
            }
        }
    }

    @Override
    public void akcja() {
        noweXY xy = new noweXY(x, y);
        ruch( xy);
        tura(xy);
    }


     
}

File: SimulationCpp/projektPO/zwierze.cpp
#include "zwierze.h"
#include <iostream>
using namespace std;

void Zwierze::tura( int nX,int nY) {
	if (!s->tab[nX][nY]) {
		s->tab[nX][nY] = s->tab[x][y];
		s->tab[x][y] = nullptr;
		x = nX;
		y = nY;
	}
	else {
		if (s->tab[nX][nY]->wypiszNazwe() == wypiszNazwe()) {
			if (plodnosc == PLODNY && s->tab[nX][nY]->plodnosc == PLODNY && this != s->tab[nX][nY]) {
				plodnosc = BEZPLODNY;
				s->tab[nX][nY]->plodnosc = BEZPLODNY;
				Organizm* n = dziecko();
				reprodukcja( nX, nY, n);
			}
		}
		else {
			s->tab[nX][nY]->kolizja(x, y);
		}
	}
}

void Zwierze::akcja() {
	int nX = x, nY = y;
	ruch( nX, nY);
	tura(nX, nY);
}

void Zwierze::ruch(int &nX,int &nY) {
	int  co, gdzie;
	co = rand() % COZMIENIA;
	gdzie = rand() % 2;
	if (gdzie == TYL) gdzie = DOTYLU;
	if (co == ZMIENIAX) {
		nX += gdzie;
	}
	else {
		nY += gdzie;
	}
	if (nY >= s->getYM() || nY < POCZATEKPLANSZY) nY -= WDRUGASTRONE * gdzie;
	if (nX >= s->getXM() || nX < POCZATEKPLANSZY) nX -= WDRUGASTRONE * gdzie;
	if (nY >= s->getYM() || nY < POCZATEKPLANSZY) nY +=  gdzie;
	if (nX >= s->getXM() || nX < POCZATEKPLANSZY) nX +=  gdzie;
}

void Zwierze::reprodukcja(int nX,int nY, Organizm* n) {
	int pX = x, pY = y;
	n->zycie = MLODY;
	n->plodnosc = NIEDOJRZALY;
	bool miejsce = false;
	for (int z = 0; z < ILOSCRODZICOW; z++) {
		if (!miejsce) {
			if (z == DRUGIRODZIC) {
				pX = nX;
				pY = nY;
			}
			for (int i =SASIADGORA; i < SASIADDOL; i++) {
			if (pX + i < POCZATEKPLANSZY || pX + i >= s->getXM())continue;
				for (int j =SASIADLEWO; j < SASIADPRAWO; j++) {
					if (pY + j <POCZATEKPLANSZY || pY + j >= s->getYM())continue;
					if (j == SRODEK && i != SRODEK || j != SRODEK && i == SRODEK) {
						if (!s->tab[pX + i][pY + j] && !miejsce) {
							miejsce = true;
							n->setX(pX + i);
							n->setY(pY + j);
							s->tab[pX + i][pY + j] = n;
							makeLog(wypiszNazwe(), wypiszNazwe(), s->log, to_string(x), to_string(y), KOMz2);
							makeLog(wypiszNazwe(), wypiszNazwe(), s->log, to_string(pX + i), to_string(pY + j), KOMz6);
						}
					}
				}
			}
		}
	}
	if (miejsce) {
		Organizm* obecny = this;
		while (obecny->getNas() != nullptr) {
			if (obecny->getNas()->getInicjatywa() < n->getInicjatywa()) break;
			obecny = obecny->getNas();
		}
		n->setPop(obecny);
		if (obecny->getNas()) {
			n->setNas(obecny->getNas());
			obecny->getNas()->setPop(n);
		}
		obecny->setNas(n);
		s->tab[nX][nY]->zycie = MLODY;
	}
	else delete[]n;
}

void Zwierze::kolizja( int nX, int nY) {
	makeLog(s->tab[nX][nY]->wypiszNazwe(), wypiszNazwe(), s->log, to_string(nX), to_string(nY), KOMz7);


	if (s->tab[nX][nY]->getSila() > sila || s->tab[nX][nY]->getSila() == sila) {
		zycie=MARTWY;
		makeLog(s->tab[nX][nY]->wypiszNazwe(), wypiszNazwe(), s->log, to_string(x), to_string(y),KOMz1);
		s->tab[x][y] = s->tab[nX][nY];
		s->tab[nX][nY] = nullptr;
		s->tab[x][y]->setX(x);
		s->tab[x][y]->setY(y);
	}
	else {
		if (s->tab[nX][nY]->getSila() < sila) {
			makeLog(wypiszNazwe(), s->tab[nX][nY]->wypiszNazwe(),  s->log, to_string(x), to_string(y),KOMz1);
			s->tab[nX][nY]->zycie =MARTWY;
			s->tab[nX][nY] = nullptr;
		}
	}
}

void Zwierze::makeLog(string z1, string z2, string& log,string x,string y, int akcja) {
	if (akcja == KOMz1)log += z1 + " zabija " + z2 + " na pozycji (" + x + "," + y + ")" + "\n";
	if (akcja == KOMz2)log += z1 + " reprodukuje " + z2 + " na pozycji (" + x + "," + y + ")" + "\n";
	if (akcja == KOMz6)log += z1 + " rodzi sie na pozycji (" + x + "," + y + ")" + "\n";
	if (akcja == KOMz3)log += z1 + " odbija " + z2 + " na pozycji (" + x + "," + y + ")" + "\n";
	if (akcja == KOMz4)log += z1 + " ucieka od " + z2 + " na pozycji (" + x + "," + y + ")" + "\n";
	if (akcja == KOMz5)log += z1 + " jest niesmiertelny nie walczy z " + z2 + " przenosi sie na pozycje (" + x + "," + y + ")" + "\n";
	if (akcja == KOMz7)log += z1 + " atakuje (" + x + "," + y + ")" + "\n";
}

Zwierze::~Zwierze() {
}